[LLDB-DAP] SBDebugger don't prefix title on progress updates

Jlalond opened the pull request:
In my last DAP patch (#123837), we piped the DAP update message into the update event. However, we had the title embedded into the update message. This makes sense for progress Start, but makes the update message look pretty wonky.

![image](https://github.com/user-attachments/assets/9f6083d1-fc50-455c-a1a7-a2f9bdaba22e)

Instead, we only use the title when it's the first message, removing the duplicate title problem.
![image](https://github.com/user-attachments/assets/ee7aefd1-1852-46f7-94bc-84b8faef6dac)

Jlalond committed:
Only include title on the first message

Jlalond committed:
Add comment explaining if and add a test

walter-erquinigo commented on lldb/include/lldb/Core/DebuggerEvents.h:
```diff
@@ -44,12 +44,16 @@ class ProgressEventData : public EventData {
-    }
-    return message;
+    // Only put the title in the message of the progress create event.
+    if (m_completed == 0) {
```
can you create some using/defines for constants like this 0? otherwise it's hard to read

clayborg commented on lldb/include/lldb/Core/DebuggerEvents.h:
```diff
@@ -44,12 +44,16 @@ class ProgressEventData : public EventData {
-      message.append(m_details);
-    }
-    return message;
+    // Only put the title in the message of the progress create event.
+    if (m_completed == 0) {
+      std::string message = m_title;
+      if (!m_details.empty()) {
+        message.append(": ");
+        message.append(m_details);
+      }
+      return message;
+    } else
+      return !m_details.empty() ? m_details : std::string();
```
We can't change the way that progress events get reported. It will work for DAP nicely, but not for anyone else. We are removing the title from any subsequent notications and that will affect everyone.

I would propose adding a new call like `SBDebugger::GetProgressFromEvent(...)` where we can get just the detail and the title as separate strings. Maybe:
```
const char *SBDebugger::GetProgressFromEvent(const lldb::SBEvent &event,
                                             uint64_t &progress_id,
                                             uint64_t &completed,
                                             uint64_t &total,
                                             bool &is_debugger_specific,
                                             const char **detail);
```
And if detail is NULL, we return the combined string of `title + ": " + detail` and if it is non NULL, the return value is the title only, and the detail gets filled into `detail`. Or we can make accessors for the title and detail like:

```
const char *SBDebugger::GetProgressTitleFromEvent(const lldb::SBEvent &event);
const char *SBDebugger::GetProgressDetailFromEvent(const lldb::SBEvent &event);
```
Any thoughts on which approach?

clayborg replied:
The current API for `const char * SBDebugger::GetProgressFromEvent(...)` is causing some memory bloat for any users as each return value, which is the `<title> + ":" + <detail>` is being constified via `ConstString` so that we can return a `const char *`. Currently I am not sure of how many `Progress` objects update the detail and return millions of updates, but if they did, that will cause use to constify all of the `<title> + ":" + <detail>` strings, which is bad. So maybe a new API should be something like:
```
bool SBDebugger::GetProgressFromEvent(
  const lldb::SBEvent &event,
  uint64_t &progress_id,
  uint64_t &completed,
  uint64_t &total,
  bool &is_debugger_specific,
  char *title, size_t max_title_length,
  char *detail, size_t max_detail_length);
```
Then we can call `strncpy` on the `Progress::m_title` and `Progress::m_detail` as needed and avoid constifying the progress strings and bloating up the const string pool.

clayborg replied:
Might be good to allow `title` and `detail` to be NULL and not filled in if they are NULL.

JDevlieghere replied:
Can we use `SBDebugger::GetProgressDataFromEvent` which returns `SBStructuredData` with all the fields and then construct the message in `lldb-dap`?

Jlalond replied:
> Can we use `SBDebugger::GetProgressDataFromEvent` which returns `SBStructuredData` with all the fields and then construct the message in `lldb-dap`?

This was going to be my alternative proposal. I assumed this wouldn't fly because I didn't know how many consumers need a workflow that doesn't follow the DAP rules. But based on the conversation I had with Greg offline the DAP progress events needs to handle it's own branching. I think the best way is returning a structured data to give us flexibility in the future.

JDevlieghere replied:
Agreed, that method was added exactly for that reason: so we didn't need to keep adding overloads to `GetProgressFromEvent` every time we added a new field. FWIW I personally think it's reasonable to construct the string in `lldb-dap`.

Jlalond committed:
Migrate to structured data

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D68927453

Jlalond commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -411,6 +412,30 @@ void SendStdOutStdErr(DAP &dap, lldb::SBProcess &process) {
     dap.SendOutput(OutputType::Stderr, llvm::StringRef(buffer, count));
 }
 
+static std::string GetStringFromStructuredData(lldb::SBStructuredData &data,
```
@JDevlieghere is there a less clunky way to do this? I know we want to handle string creation on the side of lldb-dap so we get some wins not needing to constify, but I don't like my code.

Jlalond commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -411,6 +412,30 @@ void SendStdOutStdErr(DAP &dap, lldb::SBProcess &process) {
+                                          const char *key) {
+  lldb::SBStructuredData keyValue = data.GetValueForKey(key);
+
+  if (!keyValue.IsValid())
```
@clayborg I added -1 here, I think it might just be better to let it fail at 0. That's certainly more succint, thoughts?

JDevlieghere commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -411,6 +412,30 @@ void SendStdOutStdErr(DAP &dap, lldb::SBProcess &process) {
 
+static std::string GetStringFromStructuredData(lldb::SBStructuredData &data,
+                                               const char *key) {
+  lldb::SBStructuredData keyValue = data.GetValueForKey(key);
+  if (!keyValue)
+    return std::string();
+
+  size_t size = keyValue.GetStringValue(nullptr, 0);
+  std::cout << "Size for " << key << " " << size << std::endl;
+  std::string stringValue;
+  stringValue.resize(size);
+  keyValue.GetStringValue(&stringValue[0], size + 1);
+  std::cout << "String value after: " << stringValue << std::endl;
+  return stringValue;
```
```suggestion
  lldb::SBStructuredData keyValue = data.GetValueForKey(key);
  if (!keyValue)
    return std::string();

    const size_t length = keyValue.GetStringValue(nullptr, 0);
    std::string str(length + 1, 0);
    value.GetStringValue(&str[0], length);
    return str;
```

Not really, you can avoid the resize with the ctor but otherwise that's as good as it gets (without the logging).

JDevlieghere commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -427,14 +452,27 @@ void ProgressEventThreadFunction(DAP &dap) {
+        lldb::SBStructuredData data =
+            lldb::SBDebugger::GetProgressDataFromEvent(event);
+
+        uint64_t progress_id = GetUintFromStructuredData(data, "progress_id");
+        uint64_t completed = GetUintFromStructuredData(data, "completed");
+        uint64_t total = GetUintFromStructuredData(data, "total");
```
```suggestion
        const uint64_t progress_id = GetUintFromStructuredData(data, "progress_id");
        const uint64_t completed = GetUintFromStructuredData(data, "completed");
        const uint64_t total = GetUintFromStructuredData(data, "total");
        const std::string details = GetStringFromStructuredData(data, "details");
```

Nit: I'd make these const to make it clear that we're not going to change these, as opposed to the message that we're building. I'd also move details up here so that all the message building is co-located.

JDevlieghere commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -427,14 +452,27 @@ void ProgressEventThreadFunction(DAP &dap) {
+        message += details;
+        // Verbose check, but we get -1 for the uint64 on failure to read
+        // so we check everything before broadcasting an event.
+        if (message.length() > 0 && progress_id > 0 && total >= 0 &&
+            completed >= 0)
```
`total` and `completed` are unsigned so isn't `total >= 0 && completed >= 0` always true?

Jlalond committed:
Feedback from Jonas, and cleanup dev logging

Jlalond commented:
@JDevlieghere Cleaned up as you suggested, sorry about the initial rough state but I mostly wanted a sanity check if I was overlooking anything.

clayborg reviewed (commented):
So this is changing lldb-dap progress reporting so the first reporting sends the title + detail, and subsequent entries sent just the details. How does this look in the progress API that is shown by VS Code? Does the detail just show up somewhere and remove the title?

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -428,14 +453,28 @@ void ProgressEventThreadFunction(DAP &dap) {
+        if (completed == 0) {
+          const std::string title = GetStringFromStructuredData(data, "title");
+          message += title;
+          message += ": ";
```
We should check if details contains anything before appending `": "`

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -428,14 +453,28 @@ void ProgressEventThreadFunction(DAP &dap) {
+          message += ": ";
+        }
+
+        message += details;
```
What if details is empty? We should send just the title again in that case instead of nothing? This really depends on how the progress reporting looks in VS code.

Jlalond committed:
Add new test cases, and refactor to support TITLE: DETAIL in all scenarios

Jlalond commented:
@clayborg 

DAP with a detail and a total
![image](https://github.com/user-attachments/assets/c4981158-9a19-487c-8c9b-29e55f545b37)

Indeterminate with an increment detail
![image](https://github.com/user-attachments/assets/031bcdc8-f8e5-4dde-8068-668e5dde6f5b)

Then with an indeterminate progress, but no increment detail
![image](https://github.com/user-attachments/assets/19745ee8-4204-4a88-8c64-ee705ed19355)

For number 2 I think I found another bug for indeterminate messages not updating the detail correctly, but I will address this in a future patch.

clayborg commented on lldb/test/API/tools/lldb-dap/progress/Progress_emitter.py:
```diff
@@ -37,7 +37,10 @@ def create_options(cls):
+            "--total",
+            dest="total",
+            help="Total to count up, use -1 to identify as indeterminate",
+            type="int",
```
Set the default value to `None`:
```
default=None
```

clayborg commented on lldb/test/API/tools/lldb-dap/progress/Progress_emitter.py:
```diff
@@ -37,7 +37,10 @@ def create_options(cls):
-            "--total", dest="total", help="Total to count up.", type="int"
+            "--total",
+            dest="total",
+            help="Total to count up, use -1 to identify as indeterminate",
```
Don't specify to use -1 as indeterminate, if not set, this will default to `None`, see comment below. You can say "If this option isn't specified with a valid value of 1 or higher, the progress will be indeterminate."

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -428,14 +453,35 @@ void ProgressEventThreadFunction(DAP &dap) {
+        // the title and the detail as a single message.
+        if (completed == 0 && total == 1) {
+          const std::string message =
+              GetStringFromStructuredData(data, "message");
```
is the "message" value a combined string if title + detail?

Jlalond replied:
Yes, message is title and detail

Jlalond committed:
Test refactor and comment improvements based on Greg's feedback

Jlalond committed:
Fix tests, correct for new behavior

Jlalond committed:
Python formatting

Jlalond committed:
Refactor tests to be concise and add progresse end, add additional check to sbprogress description check

clayborg commented on lldb/test/API/tools/lldb-dap/progress/Progress_emitter.py:
```diff
@@ -37,7 +37,11 @@ def create_options(cls):
-            "--total", dest="total", help="Total to count up.", type="int"
+            "--total",
+            dest="total",
+            help="Total to count up, use None to identify as indeterminate",
```
"Total items in this progress object. When this option is not specified, this will be an indeterminate progress."

clayborg commented on lldb/test/API/tools/lldb-dap/progress/TestDAP_Progress.py:
```diff
@@ -12,6 +12,41 @@
+                start_found = True
+            if "progressUpdate" in event_type:
+                message = event["body"]["message"]
+                print(f"Progress update: {message}")
```
remove print statements

clayborg commented on lldb/tools/lldb-dap/Handler/InitializeRequestHandler.cpp:
```diff
@@ -35,14 +59,47 @@ static void ProgressEventThreadFunction(DAP &dap) {
+            GetStringFromStructuredData(data, "details");
+
+        if (completed == 0) {
+          if (total == 1) {
```
Reading the internal Progress.h and Progress.cpp, this value should be reported as 
```
static constexpr uint64_t kNonDeterministicTotal = UINT64_MAX;
```
This test must not be working then or this code is never executed?

We really should add a function to `SBProgress`:
```
bool SBProgress::IsDeterministic();
```
Or  "total" should not be in the structured data if it isn't deterministic.

Jlalond replied:
Now that I think about it, I don't even think we need this `total = 1` check. We just need to check if `completed > 0;`

Jlalond replied:
I fixed this case in my last commit.

clayborg reviewed (changes requested):
Reading Progress.h/.cpp, I see that Increment will do nothing for non deterministic progress objects. We really should add a dedicated `void SBProgress::Finished()` to allow objects to say when they are done, there are no guarantees of when the object will be destroyed. So we should do that here to clean this issue up and guarantee we don't cause these tests to run longer than needed.

Jlalond committed:
Add Finalize to the test to ensure we always get the end message, and update the doc string

Jlalond committed:
Cleanup test of print statements, run formatter

Jlalond committed:
Add some basic code to the summary

clayborg commented on lldb/bindings/interface/SBProgressDocstrings.i:
```diff
@@ -11,7 +11,31 @@ The Progress class helps make sure that progress is correctly reported
+    deterministic_progress = lldb.SBProgress('Deterministic Progress', 'Detail', 3, lldb.SBDebugger)
+    for i in range(3):
+        deterministic_progress.Increment(1, f'Update {i}')
+
```
I would still call finalize for `deterministic_progress`. You could set the work items to 10 and don't send for increments as shown below for both correct deterministic usage and one that just finalized early.
```
    deterministic_progress1 = lldb.SBProgress('Deterministic Progress', 'Detail', 3, lldb.SBDebugger)
        for i in range(3):
            deterministic_progress1.Increment(1, f'Update {i}')
    # The call to Finalize() is a no-op as we already incremented the right amount of 
    # times and caused the end event to be sent.
    deterministic_progress1.Finalize() 

    deterministic_progress2 = lldb.SBProgress('Deterministic Progress', 'Detail', 10, lldb.SBDebugger)
    for i in range(3):
            deterministic_progress2.Increment(1, f'Update {i}')
    # Cause the progress end event to be sent even if we didn't increment the right 
    # number of times.
    deterministic_progress2.Finalize()

If you don't call Finalize() when the progress is not done, the progress object will 
eventually get garbage collected by the Python runtime, the end event will eventually
get sent, but it is best not to rely on the garbage collection when using
lldb.SBProgress.
```

Jlalond committed:
Fix non-deterministic title case

clayborg commented on lldb/test/API/tools/lldb-dap/progress/TestDAP_Progress.py:
```diff
@@ -4,21 +4,55 @@
+        expected_not_in_message=None,
+        only_verify_first_update=False,
+    ):
+        self.dap_server.wait_for_event("progressEnd", 5)
```
If a server is running multiple tests at once, there may be some delays. If you are sending 3 events, one second apart, I would give this timeout a larger value to ensure we don't timeout on busy machines. Something like 10 or 15 seconds at least.

Jlalond committed:
Greg comment feedback

clayborg commented on lldb/bindings/interface/SBProgressDocstrings.i:
```diff
@@ -11,7 +11,49 @@ The Progress class helps make sure that progress is correctly reported
+    deterministic_progress1 = lldb.SBProgress('Deterministic Progress', 'Detail', 3, lldb.SBDebugger) 
+    for i in range(3): 
+        deterministic_progress1.Increment(1, f'Update {i}') 
+        
```
remove empty line here

clayborg commented on lldb/bindings/interface/SBProgressDocstrings.i:
```diff
@@ -11,7 +11,49 @@ The Progress class helps make sure that progress is correctly reported
+        
+    # The call to Finalize() is a no-op as we already incremented the right amount of 
+    # times and caused the end event to be sent. 
+    deterministic_progress1.Finalize() 
```
Add empty line after this to show this goes with the code above.

clayborg commented on lldb/bindings/interface/SBProgressDocstrings.i:
```diff
@@ -11,7 +11,49 @@ The Progress class helps make sure that progress is correctly reported
+    deterministic_progress2 = lldb.SBProgress('Deterministic Progress', 'Detail', 10, lldb.SBDebugger) 
+    for i in range(3): 
+        deterministic_progress2.Increment(1, f'Update {i}') 
+        
```
Remove empty line

clayborg commented on lldb/bindings/interface/SBProgressDocstrings.i:
```diff
@@ -11,7 +11,49 @@ The Progress class helps make sure that progress is correctly reported
+    # Cause the progress end event to be sent even if we didn't increment the right 
+    # number of times. Real world examples would be in a try-finally block to ensure
+    # progress clean-up.
+
```
Remove empty line

clayborg commented on lldb/bindings/interface/SBProgressDocstrings.i:
```diff
@@ -11,7 +11,49 @@ The Progress class helps make sure that progress is correctly reported
+    # number of times. Real world examples would be in a try-finally block to ensure
+    # progress clean-up.
+
+    deterministic_progress2.Finalize() 
```
Add empty line after this.

Jlalond committed:
Add and remove lines as recommended
